firebase-installations/customer-lint-checks/src/main/java/com/google/firebase/installations/lint/IncompatibleIidVersionDetector.kt

```diff
@@ -0,0 +1,102 @@
+                IncompatibleIidVersionDetector::class.java,
+                Scope.EMPTY
+        )
+
+        @JvmField
+        val INCOMPATIBLE_IID_VERSION = Issue.create(
+                id = "IncompatibleIidVersion",
+                briefDescription = "FirebaseInstanceId version is not compatible with FirebaseInstallations",
+
+                explanation = """
+           These versions of FirebaseInstanceId and FirebaseInstallations can not be used in parallel.
+           Please update your Firebase Instance ID to a version 20.1.0 or later.
+            """,
+                category = Category.INTEROPERABILITY,
+                priority = 1,
+                severity = Severity.ERROR,
+                implementation = IMPLEMENTATION
+        )
+    }
+
+    override fun beforeCheckEachProject(context: Context) {
+        if (!context.project.isGradleProject) {
+            return
+        }
+
+        for (variant in context.project.gradleProjectModel.variants) {
+            for (lib in variant.mainArtifact.dependencies.libraries) {
+                val coordinates = lib.resolvedCoordinates
+                if (coordinates.groupId == "com.google.firebase" && coordinates.artifactId == "firebase-iid") {
+                    if (isIncompatibleVersion(coordinates)) {
+                        context.report(INCOMPATIBLE_IID_VERSION,
+                                Location.create(context.file),
+                                "Incompatible IID version found in variant ${variant.name}: ${lib.name}")
+                    }
+                }
+            }
+        }
+    }
+
+    private fun isIncompatibleVersion(coordinates: MavenCoordinates): Boolean {
+        // firebase-iid v20.1.0 already depended on FirebaseInstallations for identifier creation,
+        // but had an issue of possible silent failures. Thus, v20.1.1 or later is recommended.
+        val validIidVersionWithFis = "20.1.1"
+
+        // IID SDK version is incompatible if it is older than the IID SDK version with FIS
+        // dependency(i.e 20.1.1).
+        return compareVersions(validIidVersionWithFis, coordinates.version) == 1
+    }
+
+    private fun compareVersions(versionA: String, versionB: String): Int {
+        val versionAComponents = versionA.split('.').toTypedArray()
+        val versionBComponents = versionB.split('.').toTypedArray()
+        val k = Math.min(versionAComponents.size, versionBComponents.size)
+        for (i in 0..k - 1) {
+            if (versionAComponents.get(i) > versionBComponents.get(i)) {
+                return 1
+            }
+            if (versionAComponents.get(i) < versionBComponents.get(i)) {
+                return -1
+            }
+        }
+
+        if (versionAComponents.size == versionBComponents.size) {
+            return 0
+        }
+        if (versionAComponents.size < versionBComponents.size) {
+            return -1
+        } else {
+            return 1
+        }
+    }
```

Comment:
This method seems unnecessarily complicated.
Can't we just check for the version that we are interested in?
kind of like this (pseudo code):

versionComponents = version.split('.')
// check if correct SemVer format
return FALSE if 3 != versionComponents.size
// major version before v20? => incompatible
return FALSE if 20 > versionComponants[0].asInt
// major version after v21? => compatible
return TRUE if 21 <= versionComponants[0].asInt
// major version is v20 and minor version major version after v20.1? => compatible
return (1 <= versionComponants[1].asInt)